### PR TITLE
Rewrite "React Queued Renders" panel for better perf, and make clear it's experimental

### PR DESCRIPTION
**Pull request by markerikson**
This PR:

- supersedes the original work in #9410
  - splits the changes in that PR into two halves.  The `<ReactPanel>` changes are in this PR. The `<ReactReduxPerfPanel>` changes will go in a follow-up stacked PR.
- Scaffolds a new `packages/replay-experimental` folder by configuring TS and import sorting
- Rewrites the existing `<ReactPanel>` to improve loading perf and rendering behavior:
  - Rewrites the API calls to use `getPointStack` instead of `getAllFrames` for faster stack trace data
  - Adds logic to format point stack frames with source information
  - Adds Suspense interval caches for hits to React internal methods like `scheduleUpdateOnFiber`
  - Uses a virtualized list to render the list items
- Improves the existing logic for formatting function details:
  - Fully switches to using the `sourcesById` cache instead of Redux sources state
  - Handles formatting anonymous functions by looking at the parent function name if available (common with Redux thunks)
  - Copies support for formatting methods in classes such as `render()` from the backend routines logic
- Moves both the `<ReactPanel>` component and all associated Suspense caches over to `replay-experimental`, to clarify that this logic is not production-ready yet

image

### Future Work

There's a lot of things I've proved that I _can_ do that are not in the current version of this PR:

- Counting how many function components and class components rendered in a given render pass
- Showing list items for React render commits in the React panel (ie, alternating calls to queue a render vs commits)
- Showing the red "you are here" line in the list

### CODE REVIEW

File: packages/replay-next/src/suspense/SourcesCache.ts [hunk 1]
```diff
@@ -187,6 +188,18 @@ export const sourcesByIdCache = createSingleEntryCache<
   },
 });
 
+export const sourcesByPartialUrlCache: Cache<
```
Comment (bvaughn) [inline packages/replay-next/src/suspense/SourcesCache.ts:191]: Kind of a nit: This cache feels a bit unnecessary. Seems like we could `useMemo` or in some other way cache the result of this filter action without adding a whole new cache for it. We're really only using it to look for a single file anyway (react-dom).

File: src/ui/actions/eventListeners/eventListenerUtils.ts [hunk 2]
```diff
@@ -91,24 +99,31 @@ export type FormattedEventListener = EventListenerWithFunctionInfo & {
...
-  sourcesState: SourcesState,
+  locationInFunction: Location | MappedLocation,
   framework?: string
 ): Promise<FormattedEventListener | undefined> => {
-  const { functionLocation } = fnPreview;
-  const sources = await sourcesByIdCache.readAsync(replayClient);
-  updateMappedLocation(sources, functionLocation);
-
-  const location = getPreferredLocation(sourcesState, functionLocation);
+  // const { functionLocation } = fnPreview;
```
Comment (bvaughn) [inline src/ui/actions/eventListeners/eventListenerUtils.ts:112]:
```suggestion
```

File: packages/replay-experimental/src/components/ReactPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+import { UIThunkAction } from "ui/actions";
+import { seek } from "ui/actions/timeline";
+import { JumpToCodeButton, JumpToCodeStatus } from "ui/components/shared/JumpToCodeButton";
+import MaterialIcon from "ui/components/shared/MaterialIcon";
+import { getSourceIdsByUrl } from "ui/reducers/sources";
+import { getCurrentTime } from "ui/reducers/timeline";
+import { useAppDispatch, useAppSelector } from "ui/setup/hooks";
+
+import cardsListStyles from "ui/components/Comments/CommentCardsList.module.css";
+import styles from "ui/components/Events/Event.module.css";
```
Comment (bvaughn) [inline packages/replay-experimental/src/components/ReactPanel.tsx:27]: Seems pretty fragile for this component to import styles from other packages. We don't have TypeScript checking us there so it seems we're likely to "break" the React styles by changes made in these other CSS files.

File: src/ui/components/ReactReduxPanels.module.css [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  gap: 0.75rem;
+}
+
+.ListContainer {
+  height: 100%;
+  display: flex;
+  flex-direction: column;
+  padding: 0;
+  overflow: auto;
+}
```
Comment (bvaughn) [inline src/ui/components/ReactReduxPanels.module.css:42]: Can we delete this file? Seems like nothing is referencing it.

File: src/ui/suspense/frameCache.ts [hunk 5]
```diff
@@ -60,3 +81,99 @@ function createPauseFrames(pauseId: PauseId, frames: Frame[], sourcesState: Sour
...
+export const formattedPointStackCache: Cache<
+  [replayClient: ReplayClientInterface, point: TimeStampedPoint],
+  FormattedPointStack
+> = createCache({
+  debugLabel: "FormattedPointStack",
+  getKey: ([replayClient, point]) => point.point,
+  async load([replayClient, point]) {
+    return formatPointStackForPoint(replayClient, point);
+  },
+});
```
Comment (bvaughn) [inline src/ui/suspense/frameCache.ts:130]: Nothing referencing this cache?

```suggestion
```

File: packages/replay-experimental/src/components/ReactPanel.tsx (hunk 3 above)
Comment (markerikson) [reply packages/replay-experimental/src/components/ReactPanel.tsx:27]: Yep. This was me trying to get something that looked passably decent, as quickly as possible (and an example of why I was saying we could start extracting some reusable components or styles).

File: src/ui/components/ReactReduxPanels.module.css (hunk 4 above)
Comment (markerikson) [reply src/ui/components/ReactReduxPanels.module.css:42]: It gets used in the follow-on PR - got accidentally included here as I was trying to split up the code.

Comment (bvaughn) [reply src/ui/components/ReactReduxPanels.module.css:42]: Kill it from this PR?

File: packages/replay-experimental/src/components/ReactPanel.tsx (hunk 3 above)
Comment (bvaughn) [reply packages/replay-experimental/src/components/ReactPanel.tsx:27]: I think we should copy/fork the styles for now.

File: src/ui/suspense/frameCache.ts (hunk 5 above)
Comment (markerikson) [reply src/ui/suspense/frameCache.ts:130]: I added the cache after writing the standalone function.  The other logic should really be using the cache for consistency - I'll update that.

File: packages/replay-next/src/suspense/SourcesCache.ts (hunk 1 above)
Comment (markerikson) [reply packages/replay-next/src/suspense/SourcesCache.ts:191]: It's being used for `react-dom`, but the call is inside of a couple different caches, and I've got a spot in the follow-up PR that I just noticed ought to be using this for Redux files as well.

File: src/ui/suspense/frameCache.ts [hunk 6]
```diff
@@ -60,3 +82,108 @@ function createPauseFrames(pauseId: PauseId, frames: Frame[], sourcesState: Sour
...
+
+export async function formatPointStackForPoint(
```
Comment (bvaughn) [inline src/ui/suspense/frameCache.ts:140]: Exporting this function separately from the cache seems like a foot gun.

You can always call `cache.readAsync` to get the same behavior of the async function, but exporting the function directly seems to risk computing things that don't get cached.

Is there a reason you're exporting them separately?

It looks like the only thing that imports this function is `reactInternalsCaches.ts` and it doesn't use the import.

File: packages/replay-experimental/src/suspense/reactInternalsCaches.ts [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+  | ReactSyncUpdatedStarted
+  | ReactRenderCommitted;
+
+export const reactRendersIntervalCache = createFocusIntervalCacheForExecutionPoints<
+  [replayClient: ReplayClientInterface],
+  ReactRenderEntry
+>({
+  debugLabel: "ReactRenders",
+  getPointForValue: data => data.point.point,
+  async load(rangeStart, rangeEnd, replayClient) {
```
Comment (bvaughn) [inline packages/replay-experimental/src/suspense/reactInternalsCaches.ts:255]: This function has a lot of serial async operations. Looks like the first one is blocking, but the last few could be run in parallel.

File: src/ui/suspense/frameCache.ts [hunk 8]
```diff
@@ -60,3 +82,108 @@ function createPauseFrames(pauseId: PauseId, frames: Frame[], sourcesState: Sour
...
+export interface FormattedPointStack {
+  point: PointDescription;
+  resultPoint: PointDescription;
+  frame?: FormattedPointStackFrame;
+  allFrames: FormattedPointStackFrame[];
+  filteredFrames: FormattedPointStackFrame[];
+  functionName?: string;
+}
+
+export async function formatPointStackFrame(
```
Comment (bvaughn) [inline src/ui/suspense/frameCache.ts:104]: Same question about this function and only being used in one place. Can we either inline it into the cache or at least not export it?

File: src/ui/suspense/frameCache.ts (hunk 6 above)
Comment (markerikson) [reply src/ui/suspense/frameCache.ts:140]: In both these cases I wrote the standalone function first, then wrote the cache later, so the code had been left that way.  I can inline these.

**Pull request merged**